Use CWallet::Import* functions in all import* RPCs

achow101 opened the pull request:
#15741 introduced `ImportPrivKeys`, `ImportPubKeys`, `ImportScripts`, and `ImportScriptPubKeys` in `CWallet` which are used by `importmulti`. This PR changes the remaining `import*` RPCs (`importaddress`, `importprivkey`, `importpubkey`, and `importwallet`) to use these functions as well instead of directly adding the imported items to the wallet.

achow101 committed:
Have importprivkey use CWallet's ImportPrivKeys, ImportScripts, and ImportScriptPubKeys

Behavior changes:
* If we already have the key, it's wpkh script will still be added, although it should already be there

achow101 committed:
Have importpubkey use CWallet's ImportScriptPubKeys and ImportPubKeys functions

Behavior changes:
* If any scripts for the pubkey were already in the wallet, their timestamps will be set to 1 and label updated

achow101 committed:
Have importaddress use ImportScripts and ImportScriptPubKeys

Also removes the now unused ImportAddress and ImportScript from rpcdump.cpp

Behavior changes:
* No errors will be thrown when the script or key already exists in the wallet.
* If the key or script is already in the wallet, their labels will be updated.

achow101 committed:
Have importwallet use ImportPrivKeys and ImportScripts

Behavior changes:
* An "Importing ..." line is logged for every key, even ones that are skipped

achow101 commented:
This also makes the `CWallet` restructure simpler as the changes to `CWallet`'s import functions will then apply to all the `import*` RPCs.

achow101 committed:
Optionally allow ImportScripts to set script creation timestamp

Behavior changes:
* scripts imported in importmulti that are not explicilty scriptPubKeys will have timestamps set for them

MarcoFalke commented:
Approach ACK

promag commented on src/wallet/rpcdump.cpp:
```diff
@@ -691,31 +691,20 @@ UniValue importwallet(const JSONRPCRequest& request)
     pwallet->MarkDirty();
 
-    if (!fGood)
-        throw JSONRPCError(RPC_WALLET_ERROR, "Error adding some keys/scripts to wallet");
```
e66ed3b84593bebd3550e18b5b3be3605e08ced1

These changes are kind of breaking change, the caller has no way to know if something went bad - as it is will always succeed.Is this the case just because `dumpwallet`/`importwallet` should not be used?

achow101 replied:
This change was because it `ImportScripts` and `ImportPrivKeys` doesn't return why it failed. It could fail because the script or privkey already exists in the wallet which would still be fine, or it could cail because something went wrong with writing the data to the wallet file which is bad. I suppose I could make them return an enum, but that's a larger change that I didn't really want to do.

instagibbs replied:
The errors reported as-is appear to at least be somewhat helpful.

> It could fail because the script or privkey already exists in the wallet 

I don't think that's true. These two checks appear to be only reporting real errors. Perhaps I missed something.

achow101 replied:
Yes, but these checks are part of `ImportPrivKeys` and `ImportScripts` both of which can return false in non-error cases. Both have `HaveKey` and/or `HaveCScript` checks that can fail, but not cause the import to be bad. But there could also be an actual error which would result in the import being bad.

ryanofsky replied:
In commit "Have importwallet use ImportPrivKeys and ImportScripts" (991ecfdbbb289bd6e4c50fc63f9201edbca14f70)

> Both have HaveKey and/or HaveCScript checks that can fail, but not cause the import to be bad. 

What's this referring to? Import methods seem to return true if imported data was already found. But maybe I'm misinterpreting. Might help to refer to a specific example.

achow101 replied:
Turns out I was having a hard time with the negations.

I've pushed another commit that makes it more explicit when something is being skipped because we already have it, as well as logging when this happens.

promag reviewed (commented):
Concept ACK. Spent some time reviewing the changes commit by commit and this mostly looks good to me considering latests refactors to the wallet. IMO last commit should be reworked or add a explanation as to why it's safe to ignore errors/return values.

ryanofsky commented on src/wallet/wallet.h:
```diff
@@ -1092,7 +1092,7 @@ class CWallet final : public CCryptoKeyStore, private interfaces::Chain::Notific
     bool DummySignInput(CTxIn &tx_in, const CTxOut &txout, bool use_max_sig = false) const;
 
-    bool ImportScripts(const std::set<CScript> scripts) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
+    bool ImportScripts(const std::set<CScript> scripts, int64_t time = -1) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
```
In commit "Optionally allow ImportScripts to set script creation timestamp" (92863d7de8cf0d7432bc47f36bf1aa4de257c548)

I think it'd be better not to have a default parameter so callers will have a prompt to provide this information if it's available and not just drop it in the floor. There shouldn't be a lot of preexisting calls.

achow101 replied:
Done

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -1661,14 +1661,22 @@ bool CWallet::DummySignTx(CMutableTransaction &txNew, const std::vector<CTxOut>
             return false;
         }
+
+        if (time >= 0) {
```
In commit "Optionally allow ImportScripts to set script creation timestamp" (92863d7de8cf0d7432bc47f36bf1aa4de257c548)

This is inconsistent with existing code which uses 0 to indicate an unknown time and 1 to indicate the earliest timestamp. 

Would suggesting replacing `(time >= 0)` to `(time > 0)` here and below and changing default to 0 instead of -1. (Or dropping the default entirely).

achow101 replied:
Done

ryanofsky reviewed (commented):
Really nice changes. Almost finished reviewing (will update list below with progress).
- [X] 12fd3a849c4be0879e48e65661ca1628c21308b2 Optionally apply labels to imported scriptPubKeys (1/6)
- [X] ea665e323f76766def6c4f46a46c53e38365d179 Have importprivkey use CWallet's ImportPrivKeys, ImportScripts, and ImportScriptPubKeys (2/6)
- [X] 94ad6ad965ba6dfffaf7387b15ac659323fc85d0 Have importpubkey use CWallet's ImportScriptPubKeys and ImportPubKeys functions (3/6)
- [ ] 2fb78f8baad0778de81dc81209340e04ccef4df6 Have importaddress use ImportScripts and ImportScriptPubKeys (4/6)
- [X] 92863d7de8cf0d7432bc47f36bf1aa4de257c548 Optionally allow ImportScripts to set script creation timestamp (5/6)
- [ ] e66ed3b84593bebd3550e18b5b3be3605e08ced1 Have importwallet use ImportPrivKeys and ImportScripts (6/6)

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -175,31 +175,32 @@ UniValue importprivkey(const JSONRPCRequest& request)
+            if (pubkey.IsCompressed()) {
+                pwallet->ImportScripts({GetScriptForDestination(WitnessV0KeyHash(vchAddress))});
             }
-            pwallet->LearnAllRelatedScripts(pubkey);
```
This was done in preparation for the wallet restructure which causes `LearnAllRelatedScripts` to no longer be part of `CWallet`.

Sjors replied:
I actually find this slightly more readable than the previous `LearnAllRelatedScripts` -> `LearnRelatedScripts(key, OutputType::P2SH_SEGWIT);`. `...`. However it's hard to tell if it's really doing the same thing. Maybe add an extra commit to make that more clear?

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -175,31 +175,32 @@ UniValue importprivkey(const JSONRPCRequest& request)
             pwallet->MarkDirty();
 
+            // Use timestamp of 1 to scan the whole chain
+            if (!pwallet->ImportPrivKeys(privkey_map, 1)) {
```
Done

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -175,31 +175,32 @@ UniValue importprivkey(const JSONRPCRequest& request)
-                return NullUniValue;
+                CScript script = GetScriptForDestination(dest);
+                script_pub_keys.insert(script);
+                pwallet->ImportScriptPubKeys(strLabel, {script}, true, false, 1, (!request.params[1].isNull()) || pwallet->mapAddressBook.count(dest) == 0);
```
Done everywhere

achow101 commented on src/wallet/wallet.cpp:
```diff
@@ -1717,7 +1717,7 @@ bool CWallet::ImportPubKeys(const std::vector<CKeyID>& ordered_pubkeys, const st
 }
 
-bool CWallet::ImportScriptPubKeys(const std::string& label, const std::set<CScript>& script_pub_keys, const bool have_solving_data, const bool internal, const int64_t timestamp)
+bool CWallet::ImportScriptPubKeys(const std::string& label, const std::set<CScript>& script_pub_keys, const bool have_solving_data, const bool internal, const int64_t timestamp, bool apply_label)
```
Removed `apply_labels` and the commit that adds it.

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -511,7 +511,7 @@ UniValue importpubkey(const JSONRPCRequest& request)
 
 
-    std::string strLabel;
+    std::string strLabel = "";
```
Reverted

practicalswift commented on src/wallet/rpcdump.cpp:
```diff
@@ -699,31 +683,20 @@ UniValue importwallet(const JSONRPCRequest& request)
             const CScript& script = script_pair.first;
             int64_t time = script_pair.second;
             CScriptID id(script);
```
`id` seems to be unused?

achow101 replied:
Removed it.

Sjors commented on src/wallet/rpcdump.cpp:
```diff
@@ -175,31 +175,29 @@ UniValue importprivkey(const JSONRPCRequest& request)
+            std::set<CScript> script_pub_keys;
             for (const auto& dest : GetAllDestinationsForKey(pubkey)) {
-                if (!request.params[1].isNull() || pwallet->mapAddressBook.count(dest) == 0) {
-                    pwallet->SetAddressBook(dest, strLabel, "receive");
```
It's not obvious to me how dropping `SetAddressBook` doesn't change behavior, e.g. we're no longer calling `NotifyAddressBookChanged`.

achow101 replied:
`SetAddressBook` happens in `ImportScriptPubKeys`

Sjors reviewed (commented):
Concept ACK. I have some questions about the first commit, will study the rest later.

instagibbs commented on src/wallet/rpcdump.cpp:
```diff
@@ -543,11 +543,20 @@ UniValue importpubkey(const JSONRPCRequest& request)
-            ImportAddress(pwallet, dest, strLabel);
+            script_pub_keys.insert(GetScriptForDestination(dest));
         }
-        ImportScript(pwallet, GetScriptForRawPubKey(pubKey), strLabel, false);
```
more specifically here's the rough logic of `ImportScript`:

```
mark dirty 

addwatchonly

if is redeem script:
  addcscript
  ImportAddress
else
  SetAddressBook
```

And in this case it's not a redeemscript(last arg is false), so it marks dirty, adds watchonly, and sets address book. 

Mark dirty is top-level now, `ImportScriptPubKeys` takes care of the `SetAddressBookWithDB`, and both `ImportScriptPubKeys` and `ImportPubKeys` adds watch only in specific situations(when pubkey is not already in wallet, or  the script is not in the wallet yet). Previously the watchonly script was added when ISMINE was not spendable. I'll leave it up to other reviewers to figure if this is desired/same semantics.

achow101 replied:
I've added a note to each commit indicating what behavior changes in it.

Here, the only behavior change is that the timestamp for the key will be set to 1 if we already have it.

instagibbs commented on src/wallet/rpcdump.cpp:
```diff
@@ -540,11 +540,16 @@ UniValue importpubkey(const JSONRPCRequest& request)
+
+        pwallet->ImportScriptPubKeys(strLabel, script_pub_keys, true /* have_solving_data */, false /*internal */, 1 /* timestamp */);
+
+        pwallet->ImportPubKeys({pubKey.GetID()}, {{pubKey.GetID(), pubKey}}, {}, false /* add_keypool */, false /* internal */, 1 /* timestamp */);
```
can you annotate `key_origins` too?

achow101 replied:
Done

instagibbs commented on src/wallet/wallet.h:
```diff
@@ -1092,7 +1092,7 @@ class CWallet final : public CCryptoKeyStore, private interfaces::Chain::Notific
     bool DummySignInput(CTxIn &tx_in, const CTxOut &txout, bool use_max_sig = false) const;
 
-    bool ImportScripts(const std::set<CScript> scripts) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
+    bool ImportScripts(const std::set<CScript> scripts, int64_t time) EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
```
nit: `timestamp` since you're calling it that in other places and it matches better other args

achow101 replied:
Done

instagibbs commented on src/wallet/rpcdump.cpp:
```diff
@@ -663,51 +662,40 @@ UniValue importwallet(const JSONRPCRequest& request)
-                pwallet->WalletLogPrintf("Error importing script %s\n", HexStr(script));
-                fGood = false;
+
+            if (pwallet->ImportScripts({script}, time)) {
```
isn't this backwards?

achow101 replied:
How so?

instagibbs replied:
Because it's reporting errors when it returns true?

achow101 replied:
Oh, whoops. Fixed.

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -175,31 +175,29 @@ UniValue importprivkey(const JSONRPCRequest& request)
         {
             pwallet->MarkDirty();
 
+            // Use timestamp of 1 to scan the whole chain
```
Dunno, it's just adapted from the changed code.

jonasschnelli replied:
1 was introduced in #7551...

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -175,31 +175,29 @@ UniValue importprivkey(const JSONRPCRequest& request)
-
-            if (!pwallet->AddKeyPubKey(key, pubkey)) {
-                throw JSONRPCError(RPC_WALLET_ERROR, "Error adding key to wallet");
+            // Add the wpkh script for this key if possible
```
`ImportScriptPubKeys` doesn't actually add the scripts to the wallet, so we add the script here.

instagibbs replied:
Might be useful to say as much

achow101 committed:
Change ImportScriptPubKeys' internal to apply_label

The internal bool was only to indicate whether the given label should
be applied as things that are internal should not have a label. To make
this clearer, we change internal to apply_label and invert its usage
so things that have labels set this to true in order to have their labels
applied.

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -175,31 +175,29 @@ UniValue importprivkey(const JSONRPCRequest& request)
-                }
+                CScript script = GetScriptForDestination(dest);
+                script_pub_keys.insert(script);
+                pwallet->ImportScriptPubKeys(strLabel, {script}, true /* have_solving_data */, (request.params[1].isNull()) && pwallet->mapAddressBook.count(dest) != 0 /* internal */ , 1 /* timestamp */ );
```
I've renamed the flag to `apply_label`.

Sjors commented:
Travis sad
<img width="1212" alt="Schermafbeelding 2019-07-12 om 15 11 21" src="https://user-images.githubusercontent.com/10217/61134332-54c9ea00-a4b7-11e9-8213-8202e004a94c.png">

achow101 commented:
Fixed travis hopefully.

instagibbs commented:
utACK 991ecfdbbb289bd6e4c50fc63f9201edbca14f70

ryanofsky commented on src/wallet/rpcdump.cpp:
```diff
@@ -1258,7 +1258,7 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
         // All good, time to import
         pwallet->MarkDirty();
-        if (!pwallet->ImportScripts(import_data.import_scripts)) {
+        if (!pwallet->ImportScripts(import_data.import_scripts, timestamp)) {
```
In commit "Optionally allow ImportScripts to set script creation timestamp" (ace5d082034cbaa52f596689433381a2226a21dc)

Can you clairfy in the commit message whether this is a change in behavior for the `importmulti` RPC. Seems like a minor bugfix?

achow101 replied:
Done.

ryanofsky commented on src/wallet/rpcdump.cpp:
```diff
@@ -663,51 +662,40 @@ UniValue importwallet(const JSONRPCRequest& request)
+            std::map<CKeyID, CKey> privkey_map;
+            privkey_map.emplace(keyid, key);
+
+            if (!pwallet->ImportPrivKeys(privkey_map, time)) {
```
In commit "Have importwallet use ImportPrivKeys and ImportScripts" (991ecfdbbb289bd6e4c50fc63f9201edbca14f70)

IMO, would be clearer to call `ImportPrivKeys({keyid, key}}, time)` than have the extra map variable.

achow101 replied:
Done

ryanofsky commented on src/wallet/rpcdump.cpp:
```diff
@@ -663,51 +662,40 @@ UniValue importwallet(const JSONRPCRequest& request)
+            privkey_map.emplace(keyid, key);
+
+            if (!pwallet->ImportPrivKeys(privkey_map, time)) {
+                pwallet->WalletLogPrintf("Skipping import of %s (key already present or an error occurred)\n", EncodeDestination(PKHash(keyid)));
```
In commit "Have importwallet use ImportPrivKeys and ImportScripts" (991ecfdbbb289bd6e4c50fc63f9201edbca14f70)

Ths doesn't seem right, if HaveKey is true, ImportPrivKeys returns true.

achow101 replied:
No?

achow101 replied:
I've restored `fGood` and the error messages.

ryanofsky commented on src/wallet/rpcdump.cpp:
```diff
@@ -663,51 +662,40 @@ UniValue importwallet(const JSONRPCRequest& request)
                 continue;
             }
+
             pwallet->WalletLogPrintf("Importing %s...\n", EncodeDestination(PKHash(keyid)));
```
In commit "Have importwallet use ImportPrivKeys and ImportScripts" (991ecfdbbb289bd6e4c50fc63f9201edbca14f70)

This print statement should precede the import above...

achow101 replied:
Fixed

ryanofsky reviewed (commented):
I reviewed the whole PR, and I think it's close, but am unsure about some issues. Details are in code comments, but my main concerns were:

- It seems bad that importprivkey might be calling ImportScriptPubKeys instead of SetAddressBook just to set the label.
- There might be some unintended changes around updating nCreateTime. In general, commits are unclear about how and whether they are intended to change behavior.
- Error handling in the importwallet commit seems wrong

But aside from these specific concerns, the change seem very good and moves in a welcome direction.

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -175,31 +175,29 @@ UniValue importprivkey(const JSONRPCRequest& request)
-                }
+                CScript script = GetScriptForDestination(dest);
+                script_pub_keys.insert(script);
+                pwallet->ImportScriptPubKeys(strLabel, {script}, true /* have_solving_data */, (request.params[1].isNull()) && pwallet->mapAddressBook.count(dest) != 0 /* internal */ , 1 /* timestamp */ );
```
It won't ever add the scripts as watch only because we import the private keys earlier.

achow101 replied:
Reverted this change

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -345,10 +309,27 @@ UniValue importaddress(const JSONRPCRequest& request)
+            CScript redeem_script(data.begin(), data.end());
+
+            std::set<CScript> scripts = {redeem_script};
+            pwallet->ImportScripts(scripts);
```
I don't see the benefit of doing that.

achow101 committed:
Log when an import is being skipped because we already have it

Behavior Changes:
* Those pubkeys being imported with add_keypool set and are already in the wallet will no longer be added to the keypool

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -175,31 +175,29 @@ UniValue importprivkey(const JSONRPCRequest& request)
+
             // We don't know which corresponding address will be used;
             // label all new addresses, and label existing addresses if a
             // label was passed.
```
Reverted the below change entirely.

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -175,31 +175,29 @@ UniValue importprivkey(const JSONRPCRequest& request)
-                }
+                CScript script = GetScriptForDestination(dest);
+                script_pub_keys.insert(script);
+                pwallet->ImportScriptPubKeys(strLabel, {script}, true /* have_solving_data */, (!request.params[1].isNull()) || pwallet->mapAddressBook.count(dest) == 0 /* apply_label */ , 1 /* timestamp */ );
```
Reverted this change

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -175,31 +175,29 @@ UniValue importprivkey(const JSONRPCRequest& request)
             }
 
-            // Don't throw error in case a key is already there
-            if (pwallet->HaveKey(vchAddress)) {
```
Fixed `ImportPrivKeys`

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -345,10 +309,27 @@ UniValue importaddress(const JSONRPCRequest& request)
+            pwallet->ImportScriptPubKeys(strLabel, {script}, false /* have_solving_data */, true /* apply_label */, 1 /* timestamp */);
         } else if (IsHex(request.params[0].get_str())) {
             std::vector<unsigned char> data(ParseHex(request.params[0].get_str()));
-            ImportScript(pwallet, CScript(data.begin(), data.end()), strLabel, fP2SH);
```
To be consistent, I've also removed the fail if `HaveKey` above. This results in labels being updated if things already existed.

promag commented on src/wallet/rpcdump.cpp:
```diff
@@ -656,43 +656,34 @@ UniValue importwallet(const JSONRPCRequest& request)
-                continue;
-            }
+
             pwallet->WalletLogPrintf("Importing %s...\n", EncodeDestination(PKHash(keyid)));
```
7a58ad5624b2ea9c2d2c661446cc6e046a222fca

nit, this wouldn't be logged if the key is present.

achow101 replied:
Added a note in the commit message that this happens.

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -656,43 +656,34 @@ UniValue importwallet(const JSONRPCRequest& request)
             }
-            if (time > 0) {
-                pwallet->m_script_metadata[id].nCreateTime = time;
-                nTimeBegin = std::min(nTimeBegin, time);
```
Added the `nTimeBegin` update back.

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -1703,7 +1713,12 @@ bool CWallet::ImportPubKeys(const std::vector<CKeyID>& ordered_pubkeys, const st
+        if (GetPubKey(id, temp)) {
+            // Already have pubkey, skipping
+            WalletLogPrintf("Already have pubkey %s, skipping\n", HexStr(temp));
+            continue;
```
In commit "Log when an import is being skipped because we already have it" (8bd426d238a526f8cfa4e8dd567fb3ccadb40cf7)

Noting another change in behavior: This will now avoid adding the key to the keypool if add_keypool is true and the public key already exists. Could potentially be documented in the commit message or add_keypool description.

achow101 replied:
Added a note in the commit message

ryanofsky reviewed (approved):
utACK 35805ce74a780674c2a04259f98c8888906d7e91. Thanks for the previous fixes, and feel free to ignore my questions / comments, but I hope one you'll consider is restoring the direct SetAddressBook call in importwallet instead of calling it indirectly through ImportScriptPubKeys, because it's confusing how ImportScriptPubKeys appears to add watch only scripts.

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -656,43 +656,37 @@ UniValue importwallet(const JSONRPCRequest& request)
-            if (has_label)
-                pwallet->SetAddressBook(PKHash(keyid), label, "receive");
+
+            pwallet->ImportScriptPubKeys(label, {GetScriptForDestination(PKHash(keyid))}, true /* have_solving_data */, has_label /* apply_label */, time /* timestamp */);
```
Reverted.

achow101 commented on src/wallet/rpcdump.cpp:
```diff
@@ -656,43 +656,37 @@ UniValue importwallet(const JSONRPCRequest& request)
                 fGood = false;
                 continue;
             }
             if (time > 0) {
```
I think the use of `0` and `1` is kind of inconsistent and all of the the timestamp stuff needs to be cleaned up. But that should be done in a followup PR.

ryanofsky reviewed (approved):
utACK bc8ecc056ffa42c758b04d5bcae32c1354c3fbe1. Only change since last review was reverting the importwallet ImportScriptPubKeys change I felt was confusing, and updating one of the commit log messages to note a change in behavior.

Really nice job with this PR, and thanks for being so responsive!

instagibbs commented:
utACK https://github.com/bitcoin/bitcoin/pull/16301/commits/bc8ecc056ffa42c758b04d5bcae32c1354c3fbe1

Checked diff from https://github.com/bitcoin/bitcoin/commit/991ecfdbbb289bd6e4c50fc63f9201edbca14f70 , was reversions that were discussed and some logging/error reporting improvements.

MarcoFalke commented:
partial ACK up to af6dced944

Some questions:
* In commit af6dced944, there seems to be a behavior change that is not
  mentioned in the commit body: Previously, if the privkey already existed,
  we'd error out. Now, we'd overwrite the label?

* In commit af6dced944. Previously, there was a call to
  AddCScript(GetScriptForDestination(WitnessV0KeyHash(pubKey.GetId())));
  (via LearnAllRelatedScripts). I don't see why this is no longer required
  or where an equivalent call is made. Mind to explain?

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

partial ACK up to af6dced944

Some questions:
* In commit af6dced944, there seems to be a behavior change that is not
  mentioned in the commit body: Previously, if the privkey already existed,
  we'd error out. Now, we'd overwrite the label?

* In commit af6dced944. Previously, there was a call to
  AddCScript(GetScriptForDestination(WitnessV0KeyHash(pubKey.GetId())));
  (via LearnAllRelatedScripts). I don't see why this is no longer required
  or where an equivalent call is made. Mind to explain?
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUiFywv+Oc9kV/FPwgRVxvIEVgUQpEE4d/DcPiZ205RzdTEOXh7DxzootgnwZSzC
S8obDJGqDy26JSpCr4R6JevUl2gTYVeKMph9G2hlEHZW4pdiAxAY4/xQ8KvoAwBZ
b+u2ovlq6A8uTCQzPYkJ1KAZHHKVwd32PAYllJshilLpdVo5rqXbDwoR322/XoV2
mEeKCjnHBhQDRY4wFQ5g+2onN1oqf2Rw9fELtXqbBR7SM5EjTF79rTfkwLWmDReu
iK92/AQHfQGSrdbILz9VJ9gJE+OCco7dvi/sqMQmXKk9q3aNAY+zDgnQRvvbKkUX
p/yZrBXH8ExN7Fd89D8sb6L/h/floJM0EMQ8Ccc7cVEou77VHLC08NRiLTP5lTzH
nmmd7Br4LqZOQuVPZ9iZCRbUpu+jGPbM0uIqtYDeqO+elSnepWYTwZlCVEfjCwpX
RSaimg4I08v0cpWicoVQcFAB3DxpU0OCJZZEy4/rS5ZwkxVmyorFSy6+Hvh97oCJ
964kzylV
=2JyL
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `14e6ca5a72d68fc23e91e6eecdf4294ca99ad0e7ce1478ca529dfbbda4ec2d21  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010814e6ca5a72d68fc23e91e6eecdf4294ca99ad0e7ce1478ca529dfbbda4ec2d21f01085223bda921e7fa9c0636535469853e308fff010255f28c7ae1ca633fafa8289c55bf6a308f1045d385a83f0084b7a56bc8156ac6d0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01056369d3f01bc0788d0d690c0d22e7aa208f0208c50c64a97d95c8a0014e7fe7667a0bedd26be62df2765762df0efbacfab2fce08f1045d385a84f008e9d6172419665d990083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010bb346891011321f051a2976c5a31f2b408f1045d385a84f008c79e78c1f97ff31d0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6df010ed7b66c19c538915fe9e7beba80b320008f1045d385a83f0085599d140b6d02aa00083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

achow101 commented:
>     * In commit [af6dced](https://github.com/bitcoin/bitcoin/commit/af6dced9446eca4a67429e07b3216ae7f3d08c16), there seems to be a behavior change that is not
>       mentioned in the commit body: Previously, if the privkey already existed,
>       we'd error out. Now, we'd overwrite the label?

Updated the commit message

>     * In commit [af6dced](https://github.com/bitcoin/bitcoin/commit/af6dced9446eca4a67429e07b3216ae7f3d08c16). Previously, there was a call to
>       AddCScript(GetScriptForDestination(WitnessV0KeyHash(pubKey.GetId())));
>       (via LearnAllRelatedScripts). I don't see why this is no longer required
>       or where an equivalent call is made. Mind to explain?

`ImportPubKeys` will add the public key to `mapWatchKeys`. In doing so, it calls `ImplicitlyLearnRelatedKeyScripts` which will find that script and add it to `mapScripts` in memory. Every time this key is loaded, this script will always be found and loaded into memory. The only difference is that that script is not persisted to disk.

ryanofsky reviewed (approved):
utACK 40ad2f6a58228c72c655e3061a19a63640419378. Only change since last review is a tweaked commit message (mentioning label update in importpubkey commit)

MarcoFalke commented:
ACK 40ad2f6a58228c72c655e3061a19a63640419378 (checked that behavior changes are mentioned in the commit body)

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 40ad2f6a58228c72c655e3061a19a63640419378 (checked that behavior changes are mentioned in the commit body)
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjR+gwAwyGbOiN9UTB3m8zLjyxSUA/7k1gKPVpx2H9kX9dVAuKDzZ/9UD08dks0
M4LkQ1gHA+cvEzkEyEo4sSRcPjRajGrGIAG1TmmRR+ZexTPa8lxYd+1p7NYKu1V/
mxlee39ntChoxXfwHFj9+QiSpGFAf91/C4YnFZoZ5AfktaKPe1ZXKW2LH+lSZbKl
9dP5aL5yBIasVt9tMhLswLXZhPMUNQFQeRpev9KuM0lauiV1dN4MlyuLNg7Iops/
KK9xTth5OQIXn99KLLjnxkyZUXFlu1jjUBOOXeNXa5h9x/wJateBi/8M/Yt9YYRe
I5Rn97agU44ObjEQobsNiXusHWNGF8mv5NZlbh+iibCDnQb6MJXR1DWLgSaYx8Hf
FUtY3yhQt+2oa/wkXWhOz7jLYopT2DhmwKareWgvemscwO9rsdaO02RbdiWE4lnQ
Pbf9q3NlNdGk7KlUqt6ehGlYx9sj5cHeJl0VnH7JwibonZIznbEGf4Ej24RE8L6z
bkboCr7U
=1+uy
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `bb77b117482794207771d1571ca4e73ac19a12b52c18a1c489b86f8600c76c0b  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108bb77b117482794207771d1571ca4e73ac19a12b52c18a1c489b86f8600c76c0bf01044778683d1c49730367ec433469bc42108fff010037d05b55a6eabcde9f0ad35f85fe61c08f1045d39b196f008cf721522b6719cf30083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0105231ba3ec8f39a3727718b94b71250b708f1045d39b196f00880e522477d9b77970083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0105ee05a4df0ef4f2f50838177888077fb08f1045d39b197f0082777f6573e952cb70083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010c50b676fbffa8334b7a172d83998946808f1045d39b196f0082ebad96f2c28fe400083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

MarcoFalke commented:
I've written some tests in  #16465

Sjors reviewed (approved):
ACK 40ad2f6a5. Those extra tests also pass.
